### PR TITLE
Posts: preload when on wifi

### LINKED ISSUE #5537: My Sites:Preload site content when user navigates to the my sites page
When the user navigates to the my sites page and taps on any item for the first time the item should open to pre loaded content. To respect cellular data usage we should probably only do this on wifi.

### PR DESCRIPTION
**Pull request by nheagy**
Fixes #5537

To test:
1. While on wifi, go to a site's Blog Posts screen. The data should be preloaded after selecting the site, and take no time to show the latest posts on the Blog Post screen.
2. While on 3G or LTE, there should be no preloading. Any new posts will only appear after going to the Blog Posts screen.

Needs review: @diegoreymendez

### CODE REVIEW

File: WordPress/Classes/ViewRelated/Post/AbstractPostListViewController.swift [hunk 1]
```diff
@@ -988,11 +1002,11 @@ class AbstractPostListViewController : UIViewController, WPContentSyncHelperDele
         }
     }
 
-    func keyForCurrentListStatusFilter() -> String {
+    class func keyForCurrentListStatusFilter() -> String {
         fatalError("You should implement this method in the subclass")
     }
 
-    func currentFilterIndex() -> Int {
+    class func currentFilterIndex() -> Int {
```
Comment (diegoreymendez) [inline WordPress/Classes/ViewRelated/Post/AbstractPostListViewController.swift]: If we want to keep this method as a class method, we really need to:

- Unify the instance and class `availablePostListFilters()` methods.  Having two wouldn't be good.
- Have the setter method (`setCurrentFilterIndex()`) become a class method as well.  I know this can be annoying given the current code but these two really need to match, since they're corresponding.

The alternative would be to keep it as an instance method, which in my mind wouldn't be a bad thing either, since setting a "current" something makes the most sense when an instance owns that thing.

File: WordPress/Classes/ViewRelated/Post/PostListViewController.swift [hunk 2]
```diff
@@ -553,13 +553,13 @@ class PostListViewController : AbstractPostListViewController, UIViewControllerR
 
     // MARK: - Filter Related
 
-    override func currentPostAuthorFilter() -> PostAuthorFilter {
-        if !canFilterByAuthor() {
+    override class func currentPostAuthorFilter(blog blog:Blog) -> PostAuthorFilter {
```
Comment (diegoreymendez) [inline WordPress/Classes/ViewRelated/Post/PostListViewController.swift]: Kind of the same comment as above... having a class getter and an instance setter doesn't make much sense.  Can we unify them?

File: WordPress/Classes/ViewRelated/Post/AbstractPostListViewController.swift (hunk 1 above)
Comment (diegoreymendez) [inline WordPress/Classes/ViewRelated/Post/AbstractPostListViewController.swift]: > Unify the instance and class availablePostListFilters() methods. Having two wouldn't be good.

Let me rephrase that: it's not so much about having the same method twice, as much as the class method right now feeling like a quick patch since we can't call the instance one.

Comment (diegoreymendez) [inline WordPress/Classes/ViewRelated/Post/AbstractPostListViewController.swift]: PS: my take on these is that they should mostly be instance methods regardless of the underlying storage, which IMO could very well change to an instance variable.

Comment (nheagy) [inline WordPress/Classes/ViewRelated/Post/AbstractPostListViewController.swift]: I moved these to class methods since they need to be accessed elsewhere for preloading. I didn't want to make an instance of `PostListViewController` just to access the current filter settings. 

I agree that having a bunch of class methods on a VC is ugly, and definitely agree that the setting should correspond. I'd really like to move them out of the VC entirely, as they should be in business logic somewhere, either a model or service. Thoughts on moving them to `PostService` instead?

Comment (diegoreymendez) [inline WordPress/Classes/ViewRelated/Post/AbstractPostListViewController.swift]: > I changed these to class methods since they need to be accessed elsewhere for preloading. I didn't want to make an instance of PostListViewController just to access the current filter settings.

Agreed on the principle.

However, when tracking the calls to this method I noticed the only place making such a call without having an instance of the VC is [this one](https://github.com/wordpress-mobile/WordPress-iOS/pull/5655/files#diff-2c22ee07f415c85bb9dc440e62c7c008R544).  (Correct me if I'm wrong, that was a quick search).

That place is asking for the filter to obtain the list of statuses... while it could instead just get the list of statuses directly (which can be ok as a static method).

> I agree that having a bunch of class methods on a VC is ugly, and definitely agree that the setter should correspond. I'd really like to move them out of the VC entirely, as they should be in business logic somewhere, like a model or service. Thoughts on moving them to PostService instead?

Regarding where to place these methods, I'd suggest to keep things simple for the purpose of your PR, to try and reduce the amount of methods that were changed to `class`, and also make sure that getters and setters are on the same page too.

Having the interaction with `NSUserDefaults` in a different class could be a good idea, but I'd suggest not trying to bring that into this PR, as it would be bringing it out of it's scope a bit too much.

**Comment by diegoreymendez** [comment]
@nheagy - Since you're going to be away for a bit now, I'll assign this back to you, until you're back.  Let me know. :)

File: WordPress/Classes/ViewRelated/Blog/BlogDetailsViewController.m [hunk 3]
```diff
@@ -514,17 +514,58 @@ - (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInte
...
         self.statsService = [[WPStatsService alloc] initWithSiteId:self.blog.siteID siteTimeZone:[self.blogService timeZoneForBlog:self.blog] oauth2Token:oauthToken andCacheExpirationInterval:5 * 60];
         [self.statsService retrieveInsightsStatsWithAllTimeStatsCompletionHandler:nil insightsCompletionHandler:nil todaySummaryCompletionHandler:nil latestPostSummaryCompletionHandler:nil commentsAuthorCompletionHandler:nil commentsPostsCompletionHandler:nil tagsCategoriesCompletionHandler:nil followersDotComCompletionHandler:nil followersEmailCompletionHandler:nil publicizeCompletionHandler:nil streakCompletionHandler:nil progressBlock:nil andOverallCompletionHandler:nil];
     }
+}
+
+- (void)preloadPosts
+{
+    NSManagedObjectContext *context = [[ContextManager sharedInstance] mainContext];
+    PostService *postService = [[PostService alloc] initWithManagedObjectContext:context];
+    PostListFilterSettings *filterSettings = [[PostListFilterSettings alloc] initWithBlog:self.blog postType:PostServiceTypePost];
```
Comment (nheagy) [inline WordPress/Classes/ViewRelated/Blog/BlogDetailsViewController.m]: This is really the key reason for creating `PostListFilterSettings`. Now there are no class methods on the view controllers.

File: WordPress/Classes/Services/PostService.m [hunk 4]
```diff
@@ -700,6 +697,23 @@ - (NSDictionary *)dictionaryWithKey:(NSString *)key inMetadata:(NSArray *)metada
...
+        case PostServiceTypePost:
+            key = @"post";
+            break;
+        case PostServiceTypeAny:
+        default:
+            key = @"any";
+            break;
+    }
+    return key;
+}
```
Comment (nheagy) [inline WordPress/Classes/Services/PostService.m]: I think I'll rewrite `PostService` in swift next, and then `+ keyForType:` can be removed, and handled by the enum directly 🎉

**Comment by nheagy** [comment]
Ok, @diegoreymendez, what do you think of all _this_ stuff? 😜

File: WordPress/Classes/ViewRelated/Post/PostListFilterSettings.swift [hunk 5]
```diff
@@ -0,0 +1,166 @@
+import Foundation
+import WordPressComAnalytics
+
+@objc class PostListFilterSettings: NSObject {
```
Comment (diegoreymendez) [inline WordPress/Classes/ViewRelated/Post/PostListFilterSettings.swift]: `@objc` is only necessary if your class does not inherit from `NSObject`.  We should be able to safely remove the `@objc` here.

Comment (diegoreymendez) [inline WordPress/Classes/ViewRelated/Post/PostListFilterSettings.swift]: Also: can you add documentation to the class to explain why it exists / what it does?

File: WordPress/Classes/ViewRelated/Post/PostListFilterSettings.swift [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+    private static let currentPostAuthorFilterKey = "CurrentPostAuthorFilterKey"
+    private static let currentPageListStatusFilterKey = "CurrentPageListStatusFilterKey"
+    private static let currentPostListStatusFilterKey = "CurrentPostListStatusFilterKey"
+
+    let blog:Blog
```
Comment (diegoreymendez) [inline WordPress/Classes/ViewRelated/Post/PostListFilterSettings.swift]: Add a space after the colon (since this is how we do it throughout most of our codebase).

`let blog: Blog`

Same thing for the 2 lines below.

File: WordPress/Classes/ViewRelated/Post/PostListFilterSettings.swift [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+    let postType:PostServiceType
+    private var allPostListFilters:[PostListFilter]?
+
+    enum AuthorFilter : UInt {
+        case Mine = 0
+        case Everyone = 1
+    }
+
+    init(blog:Blog, postType:PostServiceType) {
```
Comment (diegoreymendez) [inline WordPress/Classes/ViewRelated/Post/PostListFilterSettings.swift]: Space after the colons.

File: WordPress/Classes/Services/PostService.h [hunk 8]
```diff
@@ -48,7 +52,7 @@ typedef void(^PostServiceSyncFailure)(NSError *error);
  @param success A success block
  @param failure A failure block
  */
-- (void)syncPostsOfType:(NSString *)postType
+- (void)syncPostsOfType:(PostServiceType)postType
```
Comment (diegoreymendez) [inline WordPress/Classes/Services/PostService.h]: `PostServiceType` sounds like a bad naming choice here, considering the caller is referring to the post type here, not the service type.

I understand that this method internally picks the service base on the post type, but the name should still be `PostType`, as the caller is only providing post information and is unaware of services being used internally by this class.

**Comment by diegoreymendez** [comment]
@nheagy - Like the work you did here.  Only pointed out some non-logical elements.  Let me know when ready.

File: WordPress/Classes/Services/PostService.h (hunk 8 above)
Comment (nheagy) [inline WordPress/Classes/Services/PostService.h]: Great point.

Thanks for the review, I'll have revisions for you soon 😀

**Comment by nheagy** [comment]
Alright, @diegoreymendez I made your suggested changes to `PostListFilterSettings.swift` 👍.

I ended up not renaming `PostServiceType` to `PostType` because that name is already in use, which left me without any better ideas. If you've got a suggestion, let me know :D

**Comment by diegoreymendez** [comment]
:shipit:
